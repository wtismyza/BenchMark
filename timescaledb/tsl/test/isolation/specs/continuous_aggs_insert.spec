# This file and its contents are licensed under the Timescale License.
# Please see the included NOTICE for copyright information and
# LICENSE-TIMESCALE for a copy of the license.

setup
{
    SELECT _timescaledb_internal.stop_background_workers();
    CREATE TABLE ts_continuous_test(time INTEGER, location INTEGER);
    SELECT create_hypertable('ts_continuous_test', 'time', chunk_time_interval => 10);
    CREATE OR REPLACE FUNCTION integer_now_test() returns INT LANGUAGE SQL STABLE as $$ SELECT coalesce(max(time), 0) FROM ts_continuous_test $$;
    SELECT set_integer_now_func('ts_continuous_test', 'integer_now_test');
    INSERT INTO ts_continuous_test SELECT i, i FROM
        (SELECT generate_series(0, 29) AS i) AS i;
    CREATE MATERIALIZED VIEW continuous_view
        WITH ( timescaledb.continuous, timescaledb.materialized_only=true)
        AS SELECT time_bucket('5', time), COUNT(location)
            FROM ts_continuous_test
            GROUP BY 1 WITH NO DATA;

    CREATE OR REPLACE FUNCTION lock_cagg(cagg name) RETURNS void AS $$
    DECLARE
      mattable text;
    BEGIN
      SELECT format('%I.%I', user_view_schema, user_view_name)
      FROM _timescaledb_catalog.continuous_agg
      INTO mattable;
      EXECUTE format('LOCK table %s IN EXCLUSIVE MODE', mattable);
    END; $$ LANGUAGE plpgsql;
}

teardown {
    DROP TABLE ts_continuous_test CASCADE;
}

session "I"
step "Ib"	{ BEGIN; SET LOCAL lock_timeout = '500ms'; SET LOCAL deadlock_timeout = '10ms';}
step "I1"	{ INSERT INTO ts_continuous_test VALUES (1, 1); }
step "Ic"	{ COMMIT; }

session "Ip"
step "Ipb"	{ BEGIN; SET LOCAL lock_timeout = '500ms'; SET LOCAL deadlock_timeout = '10ms';}
step "Ip1"	{ INSERT INTO ts_continuous_test VALUES (29, 29); }
step "Ipc"	{ COMMIT; }

session "S"
step "Sb"	{ BEGIN; SET LOCAL lock_timeout = '500ms'; SET LOCAL deadlock_timeout = '10ms';}
step "S1"	{ SELECT count(*) FROM ts_continuous_test; }
step "Sc"	{ COMMIT; }

session "SV"
step "SV1"	{ SELECT * FROM continuous_view order by 1; }

session "R"
setup { SET client_min_messages TO LOG; }
step "Refresh"	{ CALL refresh_continuous_aggregate('continuous_view', NULL, 15); }

session "R1"
setup { SET client_min_messages TO LOG; }
step "Refresh1"	{ CALL refresh_continuous_aggregate('continuous_view', NULL, 15); }

session "R2"
setup { SET lock_timeout = '500ms'; SET deadlock_timeout = '10ms'; }
step "Refresh2"	{ CALL refresh_continuous_aggregate('continuous_view', NULL, 15); }
teardown { SET lock_timeout TO default; SET deadlock_timeout to default; }

# the invalidation log is copied in the first materialization tranasction
session "L"
step "LockInval" { BEGIN; LOCK TABLE _timescaledb_catalog.continuous_aggs_hypertable_invalidation_log; }
step "UnlockInval" { ROLLBACK; }

# the invalidation threshold lock will block both INSERT and REFRESH
session "LI"
step "LockInvalThr" { BEGIN; LOCK TABLE _timescaledb_catalog.continuous_aggs_invalidation_threshold IN SHARE MODE; }
step "UnlockInvalThr" { ROLLBACK; }

# the invalidation threshold lock will block both INSERT and REFRESH
session "LIE"
step "LockInvalThrEx" { BEGIN; LOCK TABLE _timescaledb_catalog.continuous_aggs_invalidation_threshold ; }
step "UnlockInvalThrEx" { ROLLBACK; }

# locking a cagg's materialized hypertable will block the REFRESH in
# the second transaction, but not INSERTs.
session "LC"
step "LockCagg" { BEGIN; SELECT lock_cagg('continuous_view'); }
step "UnlockCagg" { ROLLBACK; }

# the materialization invalidation log
session "LM"
step "LockMatInval" { BEGIN; LOCK TABLE _timescaledb_catalog.continuous_aggs_materialization_invalidation_log; }
step "UnlockMatInval" { ROLLBACK; }
#only one refresh
permutation "LockInvalThrEx" "Refresh2" "Refresh"  "UnlockInvalThrEx"

#refresh and insert do not block each other once refresh is out of the
#first transaction where it moves the invalidation threshold
permutation "Ib" "LockCagg" "I1" "Refresh" "Ic" "UnlockCagg"
permutation "Ib" "LockCagg" "Refresh" "I1" "Ic" "UnlockCagg"

#refresh and select can run concurrently. Refresh blocked only by lock on
# cagg's materialized hypertable. Needs RowExclusive for 2nd txn.
permutation "Sb" "LockCagg" "Refresh" "S1" "Sc" "UnlockCagg"
permutation "Sb" "LockCagg" "S1" "Refresh" "Sc" "UnlockCagg"

#refresh will see new invalidations (you can tell since they are waiting on the invalidation log lock)
permutation "Ib" "LockInvalThr" "Refresh" "I1" "Ic" "UnlockInvalThr"
permutation "Ib" "LockInvalThr" "I1" "Refresh" "Ic" "UnlockInvalThr"

#with no invalidation threshold, inserts will not write to the invalidation log
permutation "Ib" "LockInval" "I1" "Ic" "Refresh" "UnlockInval"

#inserts beyond the invalidation will not write to the log, but since
#the refresh currently takes an exclusive lock on the invalidation
#threshold table (even if not updating the threshold) it will block
#inserts in the first transaction of the refresh
permutation "Ipb" "LockInval" "Refresh" "Ip1" "Ipc" "UnlockInval"
permutation "Ipb" "LockInval" "Ip1" "Refresh" "Ipc" "UnlockInval"
permutation "Ipb" "LockInval" "Ip1" "Ipc" "Refresh" "UnlockInval"


#refresh and insert/select do not block each other
#refresh1 is blocked on LockMatInval , insert is blocked on invalidation threshold. so refresh1 does not see the insert from I1
permutation "Refresh" "SV1" "LockMatInval" "Refresh1" "Ib" "I1" "LockInvalThrEx" "Ic" "UnlockMatInval" "UnlockInvalThrEx" "SV1"
permutation "I1" "Refresh" "LockInval" "Refresh" "Sb" "S1" "Sc" "UnlockInval"
permutation "I1" "Refresh" "LockInval" "Sb" "S1" "Refresh" "Sc" "UnlockInval"
