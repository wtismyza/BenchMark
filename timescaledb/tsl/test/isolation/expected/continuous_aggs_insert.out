Parsed test spec with 12 sessions

starting permutation: LockInvalThrEx Refresh2 Refresh UnlockInvalThrEx
step LockInvalThrEx: BEGIN; LOCK TABLE _timescaledb_catalog.continuous_aggs_invalidation_threshold ;
step Refresh2: CALL refresh_continuous_aggregate('continuous_view', NULL, 15); <waiting ...>
step Refresh: CALL refresh_continuous_aggregate('continuous_view', NULL, 15); <waiting ...>
step UnlockInvalThrEx: ROLLBACK;
step Refresh2: <... completed>
R: NOTICE:  continuous aggregate "continuous_view" is already up-to-date
step Refresh: <... completed>

starting permutation: Ib LockCagg I1 Refresh Ic UnlockCagg
step Ib: BEGIN; SET LOCAL lock_timeout = '500ms'; SET LOCAL deadlock_timeout = '10ms';
step LockCagg: BEGIN; SELECT lock_cagg('continuous_view');
lock_cagg      

               
step I1: INSERT INTO ts_continuous_test VALUES (1, 1);
step Refresh: CALL refresh_continuous_aggregate('continuous_view', NULL, 15); <waiting ...>
step Ic: COMMIT;
step UnlockCagg: ROLLBACK;
step Refresh: <... completed>

starting permutation: Ib LockCagg Refresh I1 Ic UnlockCagg
step Ib: BEGIN; SET LOCAL lock_timeout = '500ms'; SET LOCAL deadlock_timeout = '10ms';
step LockCagg: BEGIN; SELECT lock_cagg('continuous_view');
lock_cagg      

               
step Refresh: CALL refresh_continuous_aggregate('continuous_view', NULL, 15); <waiting ...>
step I1: INSERT INTO ts_continuous_test VALUES (1, 1);
step Ic: COMMIT;
step UnlockCagg: ROLLBACK;
step Refresh: <... completed>

starting permutation: Sb LockCagg Refresh S1 Sc UnlockCagg
step Sb: BEGIN; SET LOCAL lock_timeout = '500ms'; SET LOCAL deadlock_timeout = '10ms';
step LockCagg: BEGIN; SELECT lock_cagg('continuous_view');
lock_cagg      

               
step Refresh: CALL refresh_continuous_aggregate('continuous_view', NULL, 15); <waiting ...>
step S1: SELECT count(*) FROM ts_continuous_test;
count          

30             
step Sc: COMMIT;
step UnlockCagg: ROLLBACK;
step Refresh: <... completed>

starting permutation: Sb LockCagg S1 Refresh Sc UnlockCagg
step Sb: BEGIN; SET LOCAL lock_timeout = '500ms'; SET LOCAL deadlock_timeout = '10ms';
step LockCagg: BEGIN; SELECT lock_cagg('continuous_view');
lock_cagg      

               
step S1: SELECT count(*) FROM ts_continuous_test;
count          

30             
step Refresh: CALL refresh_continuous_aggregate('continuous_view', NULL, 15); <waiting ...>
step Sc: COMMIT;
step UnlockCagg: ROLLBACK;
step Refresh: <... completed>

starting permutation: Ib LockInvalThr Refresh I1 Ic UnlockInvalThr
step Ib: BEGIN; SET LOCAL lock_timeout = '500ms'; SET LOCAL deadlock_timeout = '10ms';
step LockInvalThr: BEGIN; LOCK TABLE _timescaledb_catalog.continuous_aggs_invalidation_threshold IN SHARE MODE;
step Refresh: CALL refresh_continuous_aggregate('continuous_view', NULL, 15); <waiting ...>
step I1: INSERT INTO ts_continuous_test VALUES (1, 1);
step Ic: COMMIT; <waiting ...>
step UnlockInvalThr: ROLLBACK;
step Refresh: <... completed>
step Ic: <... completed>

starting permutation: Ib LockInvalThr I1 Refresh Ic UnlockInvalThr
step Ib: BEGIN; SET LOCAL lock_timeout = '500ms'; SET LOCAL deadlock_timeout = '10ms';
step LockInvalThr: BEGIN; LOCK TABLE _timescaledb_catalog.continuous_aggs_invalidation_threshold IN SHARE MODE;
step I1: INSERT INTO ts_continuous_test VALUES (1, 1);
step Refresh: CALL refresh_continuous_aggregate('continuous_view', NULL, 15); <waiting ...>
step Ic: COMMIT; <waiting ...>
step UnlockInvalThr: ROLLBACK;
step Refresh: <... completed>
step Ic: <... completed>

starting permutation: Ib LockInval I1 Ic Refresh UnlockInval
step Ib: BEGIN; SET LOCAL lock_timeout = '500ms'; SET LOCAL deadlock_timeout = '10ms';
step LockInval: BEGIN; LOCK TABLE _timescaledb_catalog.continuous_aggs_hypertable_invalidation_log;
step I1: INSERT INTO ts_continuous_test VALUES (1, 1);
step Ic: COMMIT;
step Refresh: CALL refresh_continuous_aggregate('continuous_view', NULL, 15); <waiting ...>
step UnlockInval: ROLLBACK;
step Refresh: <... completed>

starting permutation: Ipb LockInval Refresh Ip1 Ipc UnlockInval
step Ipb: BEGIN; SET LOCAL lock_timeout = '500ms'; SET LOCAL deadlock_timeout = '10ms';
step LockInval: BEGIN; LOCK TABLE _timescaledb_catalog.continuous_aggs_hypertable_invalidation_log;
step Refresh: CALL refresh_continuous_aggregate('continuous_view', NULL, 15); <waiting ...>
step Ip1: INSERT INTO ts_continuous_test VALUES (29, 29);
step Ipc: COMMIT; <waiting ...>
step UnlockInval: ROLLBACK;
step Refresh: <... completed>
step Ipc: <... completed>

starting permutation: Ipb LockInval Ip1 Refresh Ipc UnlockInval
step Ipb: BEGIN; SET LOCAL lock_timeout = '500ms'; SET LOCAL deadlock_timeout = '10ms';
step LockInval: BEGIN; LOCK TABLE _timescaledb_catalog.continuous_aggs_hypertable_invalidation_log;
step Ip1: INSERT INTO ts_continuous_test VALUES (29, 29);
step Refresh: CALL refresh_continuous_aggregate('continuous_view', NULL, 15); <waiting ...>
step Ipc: COMMIT; <waiting ...>
step UnlockInval: ROLLBACK;
step Refresh: <... completed>
step Ipc: <... completed>

starting permutation: Ipb LockInval Ip1 Ipc Refresh UnlockInval
step Ipb: BEGIN; SET LOCAL lock_timeout = '500ms'; SET LOCAL deadlock_timeout = '10ms';
step LockInval: BEGIN; LOCK TABLE _timescaledb_catalog.continuous_aggs_hypertable_invalidation_log;
step Ip1: INSERT INTO ts_continuous_test VALUES (29, 29);
step Ipc: COMMIT;
step Refresh: CALL refresh_continuous_aggregate('continuous_view', NULL, 15); <waiting ...>
step UnlockInval: ROLLBACK;
step Refresh: <... completed>

starting permutation: Refresh SV1 LockMatInval Refresh1 Ib I1 LockInvalThrEx Ic UnlockMatInval UnlockInvalThrEx SV1
step Refresh: CALL refresh_continuous_aggregate('continuous_view', NULL, 15);
step SV1: SELECT * FROM continuous_view order by 1;
time_bucket    count          

0              5              
5              5              
10             5              
step LockMatInval: BEGIN; LOCK TABLE _timescaledb_catalog.continuous_aggs_materialization_invalidation_log;
step Refresh1: CALL refresh_continuous_aggregate('continuous_view', NULL, 15); <waiting ...>
step Ib: BEGIN; SET LOCAL lock_timeout = '500ms'; SET LOCAL deadlock_timeout = '10ms';
step I1: INSERT INTO ts_continuous_test VALUES (1, 1);
step LockInvalThrEx: BEGIN; LOCK TABLE _timescaledb_catalog.continuous_aggs_invalidation_threshold ; <waiting ...>
step Ic: COMMIT; <waiting ...>
step UnlockMatInval: ROLLBACK;
R1: NOTICE:  continuous aggregate "continuous_view" is already up-to-date
step Refresh1: <... completed>
step LockInvalThrEx: <... completed>
step UnlockInvalThrEx: ROLLBACK;
step Ic: <... completed>
step SV1: SELECT * FROM continuous_view order by 1;
time_bucket    count          

0              5              
5              5              
10             5              

starting permutation: I1 Refresh LockInval Refresh Sb S1 Sc UnlockInval
step I1: INSERT INTO ts_continuous_test VALUES (1, 1);
step Refresh: CALL refresh_continuous_aggregate('continuous_view', NULL, 15);
step LockInval: BEGIN; LOCK TABLE _timescaledb_catalog.continuous_aggs_hypertable_invalidation_log;
step Refresh: CALL refresh_continuous_aggregate('continuous_view', NULL, 15); <waiting ...>
step Sb: BEGIN; SET LOCAL lock_timeout = '500ms'; SET LOCAL deadlock_timeout = '10ms';
step S1: SELECT count(*) FROM ts_continuous_test;
count          

31             
step Sc: COMMIT;
step UnlockInval: ROLLBACK;
R: NOTICE:  continuous aggregate "continuous_view" is already up-to-date
step Refresh: <... completed>

starting permutation: I1 Refresh LockInval Sb S1 Refresh Sc UnlockInval
step I1: INSERT INTO ts_continuous_test VALUES (1, 1);
step Refresh: CALL refresh_continuous_aggregate('continuous_view', NULL, 15);
step LockInval: BEGIN; LOCK TABLE _timescaledb_catalog.continuous_aggs_hypertable_invalidation_log;
step Sb: BEGIN; SET LOCAL lock_timeout = '500ms'; SET LOCAL deadlock_timeout = '10ms';
step S1: SELECT count(*) FROM ts_continuous_test;
count          

31             
step Refresh: CALL refresh_continuous_aggregate('continuous_view', NULL, 15); <waiting ...>
step Sc: COMMIT;
step UnlockInval: ROLLBACK;
R: NOTICE:  continuous aggregate "continuous_view" is already up-to-date
step Refresh: <... completed>
